presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,271 @@
+
+        public Visitor(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types)
+        {
+            this.literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionRegistry());
+            this.metadata = metadata;
+            this.typeAnalyzer = typeAnalyzer;
+            this.session = session;
+            this.types = types;
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression, typeAnalyzer, session, types, metadata, functionInvoker, literalEncoder);
+        }
+    }
+
+    private static Expression unwrapCast(ComparisonExpression expression, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types, Metadata metadata, InterpretedFunctionInvoker functionInvoker, LiteralEncoder literalEncoder)
+    {
+        Cast cast;
+        Literal literal;
+        ComparisonExpression.Operator operator;
+
+        if (expression.getLeft() instanceof Cast && expression.getRight() instanceof Literal) {
+            cast = (Cast) expression.getLeft();
+            literal = (Literal) expression.getRight();
+            operator = expression.getOperator();
+        }
+        else if (expression.getLeft() instanceof Literal && expression.getRight() instanceof Cast) {
+            cast = (Cast) expression.getRight();
+            literal = (Literal) expression.getLeft();
+            operator = expression.getOperator().flip();
+        }
+        else {
+            return expression;
+        }
+
+        if (literal instanceof NullLiteral) {
+            return expression;
+        }
+
+        Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+        Type literalType = typeAnalyzer.getType(session, types, literal);
+
+        Optional<Type.Range> range = sourceType.getRange();
+        if (!range.isPresent()) {
+            return expression;
+        }
+
+        Object min = range.get().getMin();
+        Object max = range.get().getMax();
+
+        Object maxInTargetType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
+                session.toConnectorSession(),
+                max);
+
+        Object minInTargetType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
+                session.toConnectorSession(),
+                min);
+
+        Object literalValue = ExpressionInterpreter.evaluateConstantExpression(literal, literalType, metadata, session, ImmutableList.of());
+
+        int upperBoundComparison = compare(literalType, literalValue, maxInTargetType);
```

Comment:
I *think* it does, as long as a the types have an ordering that's consistent and the conversion is "widening" (`''` would be the minimal value in both types, so in this case, the types widen to the right). This should be the case for implicit casts. But, let me analyze this further.

Reply:
@martint when we cast from `varchar(2)` to `varchar(3)`, then yes, this should work. In case of cast to (or from?) `char(3)` saturated floor cast is actually smarter, because it can use `z`s, otherwise certain casts would not be possible to be removed.
Anyway, here i meant cast from `varchar(2)` to `integer`. It's not monotonic.